runners/core-java/src/main/java/org/apache/beam/runners/core/SimpleDoFnRunner.java

```diff
@@ -1092,15 +1099,19 @@ private void verifyAbsoluteTimeDomain() {
      * </ul>
      */
     private void setAndVerifyOutputTimestamp() {
-      // Output timestamp is currently not supported in processing time timers.
-      if (outputTimestamp != null && !TimeDomain.EVENT_TIME.equals(spec.getTimeDomain())) {
-        throw new IllegalStateException("Cannot set outputTimestamp in processing time domain.");
-      }
+
       // Output timestamp is set to the delivery time if not initialized by an user.
-      if (outputTimestamp == null) {
+      if (outputTimestamp == null && TimeDomain.EVENT_TIME.equals(spec.getTimeDomain())) {
         outputTimestamp = target;
       }
 
+      if (TimeDomain.PROCESSING_TIME.equals(spec.getTimeDomain())) {
+        outputTimestamp =
+            outputTimestampOffset.equals(Duration.ZERO)
+                ? target
+                : target.minus(offset.minus(outputTimestampOffset));
+      }
+
       if (TimeDomain.EVENT_TIME.equals(spec.getTimeDomain())) {
```

Comment:
I think that we should verify that the output timestamp is > the timestamp of the input message (if in processElement) or the output timestamp of the firing timer (if in processTimer). The < check remains correct - even for processing-time timers.

Reply:
@reuvenlax , does that mean we need to compare output timestamp with timer.currentInputWatermarkTime()? I think it will fail our previous test case of output timestamp with event timer as we have set output timestamp as 5 while having a timestamp of input element as 9.

Reply:
@reuvenlax  done